### ISSUE TITLE
SelectorParseException is broken - treats unsafe input as a String.format template.

### ISSUE THREAD

**Issue by rzwitserloot** [jhy/jsoup#1691]
This problem is a bit like the log4shell stuff / similarities with SQL injection and the like, but I don't think there is a security leak here. It is, however, a problem of trusting dirty inputs. Just, fortunately, not in a way that I can imagine can be abused for nefarious purposes.

Various lines of code in the JSoup project, for example [QueryParser.java line 47](https://github.com/jhy/jsoup/blob/master/src/main/java/org/jsoup/select/QueryParser.java#L47)  will take a string that is __very much dirty__ and could contain, well, just about anything, and passes it straight through to the [Selector.java line 168](https://github.com/jhy/jsoup/blob/master/src/main/java/org/jsoup/select/Selector.java#L168) factory method that makes a new `SelectorParseException` - this constructor does __not__ deal with dirty input.

As a consequence, in certain possibly somewhat contrived scenarios, instead of a nice clean error pointing exactly at where with a nice message telling you exactly what, you get a completely useless exception, as an exception occurs whilst constructing a nice exception. Specifically, if the raw (dirty) input contains a `%` anywhere inside it, then `String.format` will therefore attempt to treat that bit as a format specifier (such as `%5d` to indicate: a 5-digit integral number here), and crash.

One easy-enough way to trigger this is to pass invalid regexes that contain e.g. `[%&#]` - `%&` in a `String.format` template string will cause `String.format` to throw.

The only real fix is probably to review __all__ calls to line 168 in Selector.java (search for `new SelectorParseException(...)`, or better yet, in a java editor, all callers of that constructor), and make sure that the first arg definitely, guaranteed, only contains clean inputs, preferably hard constants. There are many ways to go from here, and I leave it to the core maintainers to make a decision. hence, no pullrequest.

Option 1: Review all callers and update them as follows:

```java
// Currently (QueryParser.java::47):
throw new Selector.SelectorParseException(e.getMessage());
// Should become:
throw new Selector.SelectorParseException("%s", e.getMessage());
```

Option 2: Special-case no format arguments.

```java
// Currently (Selector.java::168):
super(String.format(msg, params));
// should be:
super(params.length == 0 ? msg : String.format(msg, params));
```

Option 3: Delete this API nicety, because its dangerous. **I think this is the correct choice, but it is also the most work**.

```java
// Currently (Selector.java::167-169):
public SelectorParseException(String msg, Object... params) {
    super(String.format(msg, params));
}
// should be:
public SelectorParseException(String msg) {
    super(msg);
}
```

And on top of that, all code that now fails to compile needs to be adjusted; instead of `new SelectorParseException(a, b, c)`, change it to `new SelectorParseException(String.format(a, b, c))`, and do a quick doublecheck that the first arg definitely doesn't contain dirty strings.

## Mindset warning

Perhaps do a quick scan for `String.format` throughout the JSoup code base, this might not be the only place where raw (dirty) input ends up getting passed to places where dirty input shouldn't go.

NB: Had `String.format` had an option to do a JNDI lookup to an LDAP server.... oh dear. Then we'd have a carbon copy of the log4shell CVE. Fortunately, String.format doesn't have that functionality. Phew.

**Comment by damienhunter30** [comment]
I was thinking, the input could be scrubbed at the highest level possible, before being passed down to other methods. Just an idea, because it would be a pain in the ass to clean up and test everything.

If there's a better way to tidy this up though, would like to hear it before I attempt a fix.

**Comment by jhy** [comment]
Thanks for pointing out the formatting bug. I've added a single parameter API so that if the string doesn't need formatting (no parameters), it doesn't run through String.format. There's only six uses of the method (within the library) so it's not particularly burdensome to audit.

I find drawing a link between this and the Log4J2 / JNDI RCE is somewhat overblown. In this flow we're creating an Exception (**not** a crash) which is an IllegalStateException descender to indicate an unknown selector - or an IllegalFormatException if we tried to format that message incorrectly. There's no network path, no loading of remote classes, etc. Regardless, good to clean it up.

![image](https://user-images.githubusercontent.com/76934/147168941-2fdbd355-0325-4c34-8f29-309849c60608.png)
